File: skills/crude-daily-snippet/scripts/generate.py
#!/usr/bin/env python3
"""
原油每日价格片段生成器
用法:
  python generate.py            # 自动取最近一个完整交易日
  python generate.py 2026-06-23 # 指定评估日期

读取 ~/.platts_config.json 的 access_token，经 spgci 查询 6 个 Platts 代码，
按固定中文格式输出文字片段。
"""

import sys
import os
import json
import math
import warnings
from datetime import date, timedelta

warnings.filterwarnings('ignore')

# 确保中文在任意终端/管道下都按 UTF-8 输出
try:
    sys.stdout.reconfigure(encoding='utf-8')
except Exception:
    pass

# 代码 -> 含义
SYMBOLS = {
    "PCAAT00": "Dubai Mo1",      # 迪拜
    "PCAAV00": "Dubai Mo3",      # 迪拜 M1-M3 用
    "PCAAS00": "Dated Brent",    # Dated 现货布伦特
    "AAYES00": "ICE Brent 16:30 London",  # 伦敦16:30 ICE布伦特
    "ICLL001": "ICE Brent Settlement",    # 期货收盘 ICE布伦特
    "ICIC001": "ICE WTI Settlement",      # 期货收盘 WTI
}
# 用来确定"完整交易日"的伦敦/期货收盘锚定代码
ANCHOR = ["PCAAS00", "AAYES00", "ICLL001", "ICIC001"]


def load_token():
    token = os.environ.get('PLATTS_TOKEN')
    if token:
        return token
    path = os.path.expanduser('~/.platts_config.json')
    if os.path.exists(path):
        with open(path) as f:
            data = json.load(f)
            return data.get('access_token') or data.get('token')
    return None


def setup():
    try:
        import spgci
        import spgci.config as config
    except ImportError:
        print(json.dumps({"error": "spgci 未安装: pip install spgci --break-system-packages"}))
        sys.exit(1)
    token = load_token()
    if not token:
        print(json.dumps({"error": "未找到 Platts Token，请检查 ~/.platts_config.json"}))
        sys.exit(1)
    config.set_token(token)
    return spgci.MarketData()


def fetch_series(md, symbol, days=20):
    """返回 {date_str: value}（仅收盘 c）"""
    end = date.today()
    start = end - timedelta(days=days)
    df = md.get_assessments_by_symbol_historical(
        symbol=symbol, bate='c',
        assess_date_gte=start, assess_date_lte=end,
        page_size=500, paginate=False,
    )
    out = {}
    for _, r in df.iterrows():
        d = str(r['assessDate'])[:10]
        v = r['value']
        if v is None or (isinstance(v, float) and math.isnan(v)):
            continue
        out[d] = round(float(v), 4)
    return out


def fmt_num(x):
    """去掉多余的尾随 0"""
    s = f"{x:.4f}".rstrip('0').rstrip('.')
    return s if s not in ('', '-0') else '0'


def fmt_chg(cur, prev):
    if prev is None:
        return "—"
    diff = round(cur - prev, 4)
    if abs(diff) < 1e-9:
        return "持平"
    return ("涨" if diff > 0 else "跌") + fmt_num(abs(diff))


def val_and_prev(series, anchor):
    """返回 (anchor日值, 前一可用日值)"""
    if anchor not in series:
        return None, None
    cur = series[anchor]
    earlier = sorted(d for d in series if d < anchor)
    prev = series[earlier[-1]] if earlier else None
    return cur, prev


def main():
    md = setup()
    series = {s: fetch_series(md, s) for s in SYMBOLS}

    # 确定锚定日期
    if len(sys.argv) > 1:
        anchor = sys.argv[1]
    else:
        common = None
        for s in ANCHOR:
            ds = set(series[s].keys())
            common = ds if common is None else (common & ds)
        if not common:
            print(json.dumps({"error": "无可用的完整交易日数据"}))
            sys.exit(1)
        anchor = max(common)

    # 各代码 当日值 + 涨跌
    out = {}
    for s in SYMBOLS:
        cur, prev = val_and_prev(series[s], anchor)
        out[s] = (cur, prev)

    # 迪拜 M1-M3 价差
    t = series["PCAAT00"]
    v = series["PCAAV00"]
    common_tv = sorted(d for d in t if d in v)
    spread = {d: round(t[d] - v[d], 4) for d in common_tv}
    sp_cur, sp_prev = val_and_prev(spread, anchor)

    m = int(anchor[5:7])
    d = int(anchor[8:10])

    def line(sym):
        cur, prev = out[sym]
        if cur is None:
            return f"  （{sym} 当日无数据）"
        return f"{fmt_num(cur)}（{fmt_chg(cur, prev)}）"

    dubai = line("PCAAT00")
    dated = line("PCAAS00")
    ice_lon = line("AAYES00")
    ice_fut = line("ICLL001")
    wti_fut = line("ICIC001")
    if sp_cur is None:
        spread_line = "（无数据）"
    else:
        spread_line = f"{fmt_num(sp_cur)}（{fmt_chg(sp_cur, sp_prev)}）"

    text = (
        f"{m}月{d}日 \n"
        f"新加坡16:30\n"
        f"迪拜 {dubai}\n"
        f"迪拜M1-M3 {spread_line}\n"
        f"\n"
        f"伦敦16:30\n"
        f"Dated现货布伦特 {dated}\n"
        f"ICE布伦特 {ice_lon}\n"
        f"\n"
        f"期货收盘结算价\n"
        f"ICE布伦特 {ice_fut}\n"
        f"WTI  {wti_fut}\n"
    )
    print(text)


if __name__ == '__main__':
    main()

File: skills/crude-daily-snippet/scripts/generate.mjs
#!/usr/bin/env node

import { dirname, join } from 'path';
import { fileURLToPath } from 'url';
import { getPlattsAccessToken } from '../../../scripts/platts-auth.mjs';

const __dirname = dirname(fileURLToPath(import.meta.url));
const APPKEY = 'mXrBlqeKBqbHpYNMX96h9qN0D8H5o3AN';
const API_BASE = 'https://api.platts.com';

const SYMBOLS = {
  PCAAT00: 'Dubai Mo1',
  PCAAV00: 'Dubai Mo3',
  PCAAS00: 'Dated Brent',
  AAYES00: 'ICE Brent London 16:30',
  ICLL001: 'ICE Brent settlement',
  ICIC001: 'ICE WTI settlement'
};

const ANCHORS = ['PCAAS00', 'AAYES00', 'ICLL001', 'ICIC001'];

function isoDateInTz(date, timeZone = 'Asia/Singapore') {
  return new Intl.DateTimeFormat('en-CA', {
    timeZone,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit'
  }).format(date);
}

function addDays(isoDate, days) {
  const date = new Date(`${isoDate}T12:00:00Z`);
  date.setUTCDate(date.getUTCDate() + days);
  return date.toISOString().slice(0, 10);
}

function defaultTargetDate() {
  return addDays(isoDateInTz(new Date(), 'Asia/Singapore'), -1);
}

function fmtNum(value) {
  const n = Number(value);
  if (!Number.isFinite(n)) return null;
  const rounded = Math.round(n * 10000) / 10000;
  return rounded.toFixed(4).replace(/\.?0+$/, '') || '0';
}

function fmtChg(cur, prev) {
  if (!Number.isFinite(prev)) return '-';
  const diff = Math.round((cur - prev) * 10000) / 10000;
  if (Math.abs(diff) < 1e-9) return '持平';
  return `${diff > 0 ? '涨' : '跌'}${fmtNum(Math.abs(diff))}`;
}

function line(series, symbol, date) {
  const current = series[symbol]?.[date];
  if (!Number.isFinite(current)) return null;
  const previousDates = Object.keys(series[symbol])
    .filter(d => d < date)
    .sort();
  const previous = previousDates.length ? series[symbol][previousDates.at(-1)] : null;
  return `${fmtNum(current)}（${fmtChg(current, previous)}）`;
}

function monthDay(isoDate) {
  const [, , month, day] = isoDate.match(/^(\d{4})-(\d{2})-(\d{2})$/) || [];
  if (!month || !day) throw new Error(`Invalid date: ${isoDate}`);
  return `${Number(month)}月${Number(day)}日`;
}

async function fetchHistory(token, symbols, startDate, endDate) {
  const filter = `symbol in ("${symbols.join('","')}") AND assessDate>="${startDate}" AND assessDate<="${endDate}"`;
  const url = `${API_BASE}/market-data/v3/value/history/symbol?${new URLSearchParams({
    filter,
    pageSize: '5000'
  })}`;

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${token}`,
      appkey: APPKEY
    }
  });

  if (!response.ok) {
    const text = await response.text();
    throw new Error(`Platts history API HTTP ${response.status}: ${text.slice(0, 300)}`);
  }

  return response.json();
}

function normalizeSeries(payload) {
  const series = Object.fromEntries(Object.keys(SYMBOLS).map(symbol => [symbol, {}]));

  for (const result of payload?.results || []) {
    const symbol = result?.symbol;
    if (!series[symbol]) continue;

    for (const item of result?.data || []) {
      if (String(item?.bate || '').toLowerCase() !== 'c') continue;
      const date = String(item?.assessDate || '').slice(0, 10);
      const value = Number(item?.value);
      if (/^\d{4}-\d{2}-\d{2}$/.test(date) && Number.isFinite(value)) {
        series[symbol][date] = value;
      }
    }
  }

  return series;
}

function latestCompleteDate(series) {
  let common = null;
  for (const symbol of ANCHORS) {
    const dates = new Set(Object.keys(series[symbol] || {}));
    common = common ? new Set([...common].filter(date => dates.has(date))) : dates;
  }
  if (!common?.size) return null;
  return [...common].sort().at(-1);
}

function requireDate(series, date) {
  const missingAnchors = ANCHORS.filter(symbol => !Number.isFinite(series[symbol]?.[date]));
  if (missingAnchors.length) {
    throw new Error(`${date} incomplete: missing ${missingAnchors.join(', ')}`);
  }
}

function buildSnippet(series, targetDate) {
  const spread = {};
  for (const date of Object.keys(series.PCAAT00 || {})) {
    if (Number.isFinite(series.PCAAV00?.[date])) {
      spread[date] = series.PCAAT00[date] - series.PCAAV00[date];
    }
  }
  series.DUBAI_M1_M3 = spread;

  const required = {
    PCAAT00: line(series, 'PCAAT00', targetDate),
    DUBAI_M1_M3: line(series, 'DUBAI_M1_M3', targetDate),
    PCAAS00: line(series, 'PCAAS00', targetDate),
    AAYES00: line(series, 'AAYES00', targetDate),
    ICLL001: line(series, 'ICLL001', targetDate),
    ICIC001: line(series, 'ICIC001', targetDate)
  };

  const missing = Object.entries(required)
    .filter(([, value]) => !value)
    .map(([symbol]) => symbol);
  if (missing.length) {
    throw new Error(`${targetDate} missing values: ${missing.join(', ')}`);
  }

  return [
    `${monthDay(targetDate)} `,
    '新加坡16:30',
    `迪拜 ${required.PCAAT00}`,
    `迪拜M1-M3 ${required.DUBAI_M1_M3}`,
    '',
    '伦敦16:30',
    `Dated现货布伦特 ${required.PCAAS00}`,
    `ICE布伦特 ${required.AAYES00}`,
    '',
    '期货收盘结算价',
    `ICE布伦特 ${required.ICLL001}`,
    `WTI  ${required.ICIC001}`
  ].join('\n');
}

async function main() {
  const args = process.argv.slice(2).filter(arg => !arg.startsWith('--'));
  const targetDate = args[0] || defaultTargetDate();
  if (!/^\d{4}-\d{2}-\d{2}$/.test(targetDate)) {
    throw new Error('Usage: node generate.mjs [YYYY-MM-DD]');
  }

  const token = await getPlattsAccessToken({ quiet: true });
  const startDate = addDays(targetDate, -14);
  const endDate = addDays(targetDate, 1);
  const payload = await fetchHistory(token, Object.keys(SYMBOLS), startDate, endDate);
  const series = normalizeSeries(payload);

  if (args[0]) {
    requireDate(series, targetDate);
  } else {
    const latest = latestCompleteDate(series);
    if (!latest) throw new Error('No complete trading day found');
    if (latest !== targetDate) {
      throw new Error(`上一日 ${targetDate} 数据未齐；最新完整交易日为 ${latest}`);
    }
  }

  process.stdout.write(`${buildSnippet(series, targetDate)}\n`);
}

main().catch(error => {
  process.stderr.write(`[crude-daily-snippet] ${error.message}\n`);
  process.exit(1);
});
